lib/logflare/vault.ex

```diff
@@ -0,0 +1,90 @@
+defmodule Logflare.Vault do
+  @doc """
+  GenServer needed for Cloak.
+  It handles secrets migration for key rolling at startup.
+
+  To run the migration at runtime, use the following:
+  ```elixir
+  iex> Logflare.Vault.do_migrate()
+  ```
+  An old encryption key should be present for the migration.
+
+  """
+  use Cloak.Vault, otp_app: :logflare
+
+  alias Cloak.Ecto.Migrator
+  require Logger
+
+  @schemas [
+    Logflare.Backends.Backend
+  ]
+
+  @impl GenServer
+  def init(config) do
+    if Application.get_env(:logflare, :env) == :test do
+      # make ets table public
+      :ets.new(@table_name, [:named_table, :public])
+    end
+
+    default_key = Application.get_env(:logflare, :encryption_key_default) |> maybe_decode!()
+    retired_key = Application.get_env(:logflare, :encryption_key_retired) |> maybe_decode!()
+
+    ciphers =
+      [
+        default:
+          {Cloak.Ciphers.AES.GCM, tag: "AES.GCM.V1." <> hash(default_key), key: default_key},
+        retired:
+          if(retired_key != nil,
+            do: {Cloak.Ciphers.AES.GCM, tag: "AES.GCM.V1" <> hash(retired_key), key: retired_key},
+            else: nil
+          )
+      ]
+      |> Enum.filter(fn {_k, v} -> v != nil end)
+
+    config = Keyword.put(config, :ciphers, ciphers)
+
+    Task.start_link(fn ->
+      # wait for genserver config to be saved, see https://github.com/danielberkompas/cloak/blob/v1.1.4/lib/cloak/vault.ex#L186
+      :timer.sleep(1_000)
+
+      if Keyword.has_key?(ciphers, :retired) do
```

Comment:
is it worth having a check that `default != retired` to skip the migration if for some reason identical keys are passed? Maybe Cloak does this internally already but if not, saving cycles on not "re"-encrypting data would be nice

Reply:
the migration won't run for data that is on the default key, as cloak tags the encrypted string and filters based on it. hence why the sha256 hash is needed.